A notes sharing app in drf

hamdanasim49 opened the pull request:
Please review my code

Commit:
A notes sharing app in drf

WHAT: Implemented the version history and commenting feature

WHY: It is our assignment

IMPACT: HIGH

Commit:
A notes sharing app in drf

WHAT: Implemented the version history and commenting feature

WHY: It is our assignment

IMPACT: HIGH

Commit:
A notes sharing app in drf

WHAT: Implemented the version history and commenting feature

WHY: It is our assignment

IMPACT: HIGH

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
 
 
-class Note(models.Model):
+class AbstractClass(models.Model):
```
The name of the class should be more specific. AbstractModel would be more suited.

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
+    """
+
+    title = models.CharField(max_length=31, default=" ")
+    text = models.CharField(max_length=63)
```
for text we can use textfield, charfield is limited to to certain car length.

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
+    title = models.CharField(max_length=31, default=" ")
+    text = models.CharField(max_length=63)
     archive = models.BooleanField(default=False)
     user = models.ForeignKey(User, on_delete=models.CASCADE, null=True)
```
A note should have an author. Please remove null=True

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
+    database, it will have foreign key for user and note as well as body of content
+    """
+
+    user = models.ForeignKey(User, on_delete=models.CASCADE, null=True)
```
A comment will always have an author. Please remove null=True

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
+
+    user = models.ForeignKey(User, on_delete=models.CASCADE, null=True)
+    note = models.ForeignKey(
+        Note, on_delete=models.CASCADE, null=True, related_name="comments"
```
please state the reason why have we set null=True here?

mubashirT commented on notes/models.py:
```diff
@@ -2,21 +2,69 @@
+    note = models.ForeignKey(
+        Note, on_delete=models.CASCADE, null=True, related_name="comments"
+    )
+    text = models.CharField(max_length=127, default=" ", null=True)
```
Should we let users add an empty comment? and why is the default text set to a space?
We should use textfield here.

mubashirT commented on notes/permissions/permissions.py:
```diff
@@ -12,3 +12,9 @@ def has_permission(self, request, view):
+
+class SharedPermission(permissions.BasePermission):
+    def has_object_permission(self, request, view, obj):
+        print("Allo", obj)
```
please remove print statements.

mubashirT commented on notes/serializers/serializers.py:
```diff
@@ -1,8 +1,24 @@
+from users.serializers.serializers import UserSerializer
+
+
+class CommentOneSerializer(serializers.ModelSerializer):
```
Please rename this serializer. The name is not intuitive.

mubashirT commented on notes/serializers/serializers.py:
```diff
@@ -19,24 +37,100 @@ class Meta:
+            note_id=instance,
+            title=instance.title,
+            text=instance.text,
+            edited_by=self.context["request"].user,
```
This is not a save way to access user. Please ensure that user exists in instance before accessing it.

mubashirT commented on notes/serializers/serializers.py:
```diff
@@ -19,24 +37,100 @@ class Meta:
+        """
+
+        representation = super().to_representation(instance)
+        if representation["comments"]:
```
Please improve the logic for getting the last comment. You can improve it by just fetching the last comment in the queryset.

mubashirT commented on notes/serializers/serializers.py:
```diff
@@ -19,24 +37,100 @@ class Meta:
+
+    def paginated_comments(self, obj):
+        paginator = PageNumberPagination()
+        paginator.page_size = 10
```
please use a constant for this page size.

mubashirT commented on notes/views/views.py:
```diff
@@ -57,4 +65,50 @@ def get_queryset(self):
+    @action(detail=True, methods=["GET"], name="versions")
+    def versions(self, request, pk=None):
+        paginator = PageNumberPagination()
+        paginator.page_size = 10
```
Please use page size constant here.

mubashirT commented on notes/views/views.py:
```diff
@@ -57,4 +65,50 @@ def get_queryset(self):
+    @action(detail=True, methods=["GET"], name="comments")
+    def comments(self, request, pk=None):
+        paginator = PageNumberPagination()
+        paginator.page_size = 10
```
Looks like there is code repetition here.

mubashirT commented on notes/views/views.py:
```diff
@@ -57,4 +65,50 @@ def get_queryset(self):
+    """
+
+    queryset = Comment.objects.all()
+    authentication_classes = (JWTAuthentication,)
```
we do not need to tell the auth class here.

mubashirT commented on users/tests/tests.py:
```diff
@@ -0,0 +1,75 @@
+class RegisterTestCase(APITestCase):
+    """This class is used to test the user registration"""
+
+    data = {
```
please use faker class here to create random data every time.

mubashirT reviewed (changes requested):
There are a lot issues with the PR. Please drop everything and fix these issues.

Commit:
A notes sharing app in drf

WHAT: Implemented the version history and commenting feature

WHY: It is our assignment

IMPACT: HIGH

Commit:
A notes sharing app in drf

WHAT: Implemented the version history and commenting feature

WHY: It is our assignment

IMPACT: HIGH
